transport/x/cherami/example/server.go

```
@@ -0,0 +1,95 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package example
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/yarpc"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/transport/x/cherami"
+	"go.uber.org/yarpc/transport/x/cherami/example/thrift/example/exampleserviceserver"
+)
+
+// ServerConfig is the configuration needed to create a ExampleService
+type ServerConfig struct {
+	destination   string
+	consumerGroup string
+}
+
+// Service demos a service that implements an oneway procedure(defined in thrift/example.thrift)
+type Service struct {
+	dispatcher *yarpc.Dispatcher
+	config     ServerConfig
+}
+
+// NewService creates a example service
+func NewService(config ServerConfig) *Service {
+	return &Service{
+		config: config,
+	}
+}
+
+// Start starts the service
+func (s *Service) Start() error {
+	// frontend ip is only needed in local testing
+	// for a real production server, frontend ip can be empty and hyperbahn will be used to connect to cherami
+	frontend, err := getLocalIP()
+	if err != nil {
+		return nil
+	}
+	port := 4922
+
+	// Server side needs to start the server dispatcher and register itself which implements the procedures
+	s.dispatcher = yarpc.NewDispatcher(yarpc.Config{
+		Name: "server",
+		Inbounds: []transport.Inbound{cherami.NewInbound(cherami.InboundConfig{
+			Destination:   s.config.destination,
+			ConsumerGroup: s.config.consumerGroup,
+			Frontend:      frontend,
+			Port:          port,
+		})},
+	})
+	if err := s.dispatcher.Start(); err != nil {
+		return err
+	}
+
+	serverCalled = make(chan string)
+
+	// Now register the service to dispatcher
+	s.dispatcher.Register(exampleserviceserver.New(&Service{}))
```

@abhinav I think we should throw an error if Register is called after Start... what do you think?